Direct Boot support for Custom Ringtones

LaserKaspar opened the pull request:
Fixes https://github.com/BlackyHawky/Clock/issues/292

Tested on Xiaomi Mi 9T Pro - Evolution X ROM A14

I hope I didn't break some other functionality. Let me know if you need further testing/changes. 

Thanks for your quick responses earlier!

LaserKaspar committed:
Direct Boot support for Custom Ringtones

BlackyHawky commented on app/src/main/java/com/best/deskclock/data/RingtoneModel.java:
```diff
@@ -103,28 +110,64 @@ public RingtoneModel(Context context, SharedPreferences prefs) {
+        Uri safeUri = uri;
+
+        // On API 24+ (Android 7.0+), copy to device protected storage
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
```
Instead, use the codes indicated in the SdkUtils.java file

BlackyHawky commented on app/src/main/java/com/best/deskclock/utils/RingtoneUtils.java:
```diff
@@ -90,12 +92,21 @@ public static boolean isRingtoneUriReadable(Context context, Uri uri) {
     public static MediaPlayer createPreparedMediaPlayer(Context context, Uri... ringtoneUris) {
+        // Use a Direct Boot aware context if needed
+        Context safeContext = context;
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) { // Direct Boot is only supported since API 24 (Android 7.0+)
```
Same thing here: use the codes indicated in the SdkUtils.java file

BlackyHawky commented:
Thank you very much for working on this subject. 👍

I'll test it in a few days (not free for the next 2 days).

Edit: One thing before I forget: I've been working on replacing `MediaPlayer` with `ExoPlayer` to play ringtones.

Even if the RingtoneUtils.java file hasn't been modified in my change, I'm afraid it'll break your modification. 😬

Let's talk about it again soon.

BlackyHawky commented on app/src/main/java/com/best/deskclock/data/RingtoneModel.java:
```diff
@@ -103,28 +110,64 @@ public RingtoneModel(Context context, SharedPreferences prefs) {
         }
     }
 
+
```
Please, delete this line.

BlackyHawky commented on app/src/main/java/com/best/deskclock/data/RingtoneModel.java:
```diff
@@ -103,28 +110,64 @@ public RingtoneModel(Context context, SharedPreferences prefs) {
+
+            File destFile = new File(directBootContext.getFilesDir(), filename);
+            try (InputStream in = mContext.getContentResolver().openInputStream(uri);
+                OutputStream out = new FileOutputStream(destFile)) {
```
Please rename `in` to `inputStream` and `out` to `outPutStream`. Do the same line 115.

BlackyHawky commented:
@LaserKaspar 
I have now finished working on replacing `MediaPlayer` with `ExoPlayer`.

So can you please update your branch and retest to check that your change still works?

LaserKaspar committed:
Merge branch 'BlackyHawky:main' into main

LaserKaspar committed:
Fix: only delete files in device protected storage

LaserKaspar committed:
Update ExoPlayer to use DirectBoot aware context.

LaserKaspar commented:
Updated ExoPlayer to use DirectBoot aware context and tested again. 
Working for me now. 

One thing to consider is that if for some reason a device sound (URL provided by System Sounds - Sound & Vibration/Default alarm sound) is set to a file in user storage, it's not accessible during DirectBoot so it won't work. For now it just does a fallback to a default alarm which hopefully is accessible (didn't test).
To support that you would need to cache the sounds when creating the alarm instead of doing it when a custom alarm is added, if that's something you'd want to support we could look into that / other ways to do it.

BlackyHawky commented on app/src/main/java/com/best/deskclock/ringtone/RingtonePlayer.java:
```diff
@@ -223,6 +232,24 @@ public void play(Uri ringtoneUri, long crescendoDuration) {
 
         mExoPlayer.prepare();
     }
+    /**
```
Please, add a new line here.

BlackyHawky commented on app/src/main/java/com/best/deskclock/data/RingtoneModel.java:
```diff
@@ -118,6 +154,20 @@ void removeCustomRingtone(Uri uri) {
         final List<CustomRingtone> ringtones = getMutableCustomRingtones();
         for (CustomRingtone ringtone : ringtones) {
             if (ringtone.getUri().equals(uri)) {
+
```
Please, delete this empty line.

BlackyHawky commented:
> Updated ExoPlayer to use DirectBoot aware context and tested again. 
> Working for me now. 

Very good news 👍

> One thing to consider is that if for some reason a device sound (URL provided by System Sounds - Sound & Vibration/Default alarm sound) is set to a file in user storage, it's not accessible during DirectBoot so it won't work. For now it just does a fallback to a default alarm which hopefully is accessible (didn't test).
> To support that you would need to cache the sounds when creating the alarm instead of doing it when a custom alarm is added, if that's something you'd want to support we could look into that / other ways to do it.

- Do you mean that it is possible to customize the default alarm ringtone directly in the device settings? 🤔

- Why cache sounds if no alarms are created? (I'm thinking of users who don't use this app to set alarms but only to use the timer for example)

---

In the RingtonePlayer.java file you created the `isUriAccessible` method; can't we use the existing `RingtoneUtils.isRingtoneUriReadable` method instead?

LaserKaspar committed:
Use RingtoneUtils.isRingtoneUriReadable instead of local function.

LaserKaspar commented:
> Do you mean that it is possible to customize the default alarm ringtone directly in the device settings? 🤔
Yes, at least on my phone.

Adding sounds in: Settings -> Sound & Vibration -> Default alarm sound
<p>
<img align="left" width="40%" src="https://github.com/user-attachments/assets/1884f3bb-643e-422f-a7d2-0292e65914e8">
<img width="40%" src="https://github.com/user-attachments/assets/fd9b08b5-7dcd-426b-99ae-4950f2317ce0">
</p>
would then be made available as system sounds in this list:

And the clock app would assume it is accessible because it is provided by the system and not the user.

> Why cache sounds if no alarms are created? (I'm thinking of users who don't use this app to set alarms but only to use the timer for example)

No need to cache sounds if no alarm is created. 
For every created alarm the sound would need to be cached instead of only custom sounds (Or at least check if it is available in DirectBoot and if not cache it).

---
> In the RingtonePlayer.java file you created the isUriAccessible method; can't we use the existing RingtoneUtils.isRingtoneUriReadable method instead?

Absolutely! I just missed that the other one already exists. Thanks for pointing that out. Fixed.

BlackyHawky commented:
@LaserKaspar 
Great, the code is well written but don't forget to correct the few comments written above. 😉

Regarding the last topic, do you think it is really necessary?
Indeed, on my 2 main devices I don't have the settings you mention; I'm afraid that it will create bugs on devices that support `DirectBoot` but don't have the same device settings as you.

I think we should link to custom ringtones only in the app and not from the device settings.

LaserKaspar commented:
> ... correct the few comments written above.

What comments are you referring to exactly?

> Regarding the last topic, do you think it is really necessary?

No, just something to keep in mind / a shortcoming of the simpler solution. For my use case it's exactly what I needed.

BlackyHawky commented:
@LaserKaspar 
> What comments are you referring to exactly?

Comments in the code; this is mainly just formatting. Don't you see them when you click on _"Files changed"_ at the top of the discussion?

> No, just something to keep in mind / a shortcoming of the simpler solution. For my use case it's exactly what I needed.

If it is not necessary then let's leave it as it is now.

LaserKaspar committed:
Fixed formatting and used available version checks

LaserKaspar commented:
Ahh I see, after you "requested changes" they also appeared in the chat history. Didn't get notified / Didn't know they were there previously. 
Thanks for pointing them out. Fixed.

EDIT: Now they're gone, probably because I changed / fixed the files? Hope I didn't miss anything.

BlackyHawky commented:
@LaserKaspar 
After a quick test and after viewing your changes on Android Studio:
- Now we can add the same ringtone multiple times which is not wanted;
- The added ringtone is not immediately played which was the case before;

---

In the RingtoneModel.java file:
- You wrote:
  ```java
  if (file.getAbsolutePath().startsWith(directBootDir.getAbsolutePath())) {
      file.delete();
  }
  ```
  
  I don't think this is the correct use of `delete()`.
  You should instead write:
  
  ```java
  if (directBootDir != null && file.getAbsolutePath().startsWith(directBootDir.getAbsolutePath())) {
      boolean deleted = file.delete();
      if (!deleted) {
          LogUtils.e("Failed to delete local log file");
      }
  }
  ```

- You forgot to remove the import `android.os.Build;`
- Can you please resolve the Android Studio warnings? 
  For example, replace:
  
  ```java
  File file = new File(uri.getPath());
  ```
  
  with:
  
  ```java
  File file = new File(Objects.requireNonNull(uri.getPath()));
  ```

- You forgot to rename "in" to "inputStream" on line 115.

---

In the RingtonePlayer.java file:

-  You forgot to remove the imports;

LaserKaspar committed:
upgrading context on object creation; resolved warnings

LaserKaspar committed:
fixed initial playback after adding custom ringtone

LaserKaspar commented:
Thanks for pointing them out. I hope I'm not too much of a hassle for you! 😄 

> Now we can add the same ringtone multiple times which is not wanted;

Fixed!

> The added ringtone is not immediately played which was the case before;

Fixed, but I needed to propagate the new Uri, there might be a better way to do this. Let my know if you approve of my approach.

In DataModel.java there is still a warning, but I don't know if I caused it.

BlackyHawky committed:
Some formatting

BlackyHawky commented:
@LaserKaspar 
> Thanks for pointing them out. I hope I'm not too much of a hassle for you! 😄

Absolutely not at all! Don't worry about it. 😉

---

Otherwise, after another try:
- We can no longer add ringtones by selecting a folder, which is very problematic 😬;
- When adding multiple ringtones, the app size increases; I guess this is normal (the good thing is that if you move the ringtone to another folder, the app remembers the copied ringtone uri for DirectBoot);

LaserKaspar committed:
Fixed adding custom ringtones by folder

LaserKaspar commented:
Was quite busy this week. So here's my quite late update.

> We can no longer add ringtones by selecting a folder, which is very problematic 😬

Fixed. Didn't know it was possible, so didn't test it. Sorry.

> When adding multiple ringtones, the app size increases; I guess this is normal (the good thing is that if you move the ringtone to another folder, the app remembers the copied ringtone uri for DirectBoot);

This is indeed normal. I don't think I can do anything about that. If you don't want that we could include a setting that disables the caching and therefore Direct Boot. Although the app size should also increase when only adding a single one.

BlackyHawky committed:
Small adjustments

BlackyHawky committed:
Delete unnecessary comment

BlackyHawky commented:
@LaserKaspar 
Hi,

I've also been very busy these past few days.

The code is well written and the tests seem to be working.

If users complain about this change, can I count on you to answer their questions/comments?

Again, thank you for working on this topic. 👍

LaserKaspar commented:
> If users complain about this change, can I count on you to answer their questions/comments?

Sure, just mention me if you need something.

I also found another issue that might be related to my changes so I'll state it here instead of opening a new issue. 

1. If a default alarm-sound is selected in the apps settings, and the sound is then deleted from disk, it is still referenced by alarms that used the default instead of using the system default / an internal default and therefore alarms don't fire (which in my opinion should never happen and there should always be a fallback sound instead of playing a silent alarm).
2. If an alarm-sound is deleted and the sound is still referenced by some alarms they fallback to system default alarm sound instead of the one specified in the app.

BlackyHawky commented:
@LaserKaspar 
> 1. If a default alarm-sound is selected in the apps settings, and the sound is then deleted from disk, it is still referenced by alarms that used the default instead of using the system default / an internal default and therefore alarms don't fire (which in my opinion should never happen and there should always be a fallback sound instead of playing a silent alarm).

Indeed, I think this is due to your change; it should not happen as indicated [here](https://github.com/BlackyHawky/Clock/blob/172da7433f44cab3c34df207b211a30f9c3cdc76/app/src/main/java/com/best/deskclock/ringtone/RingtonePlayer.java#L209-L211).
It seems that the uri kept in memory is that of the file copied into the protected storage.
Is this really disturbing?
Maybe we can change the [isRingtoneUriReadable()](https://github.com/BlackyHawky/Clock/blob/172da7433f44cab3c34df207b211a30f9c3cdc76/app/src/main/java/com/best/deskclock/utils/RingtoneUtils.java#L78) method?

> 2. If an alarm-sound is deleted and the sound is still referenced by some alarms they fallback to system default alarm sound instead of the one specified in the app.

I don't understand the end of your sentence: _"instead of the one specified in the app"_ 🤔

---

Do you think you can correct the two points you mentioned? If you want, you can work on that.

LaserKaspar commented:
> Maybe we can change the [isRingtoneUriReadable()](https://github.com/BlackyHawky/Clock/blob/172da7433f44cab3c34df207b211a30f9c3cdc76/app/src/main/java/com/best/deskclock/utils/RingtoneUtils.java#L78) method?

Done, any alarm the would not be played should now correctly fallback to the default sound. (Tested on my device)

> I don't understand the end of your sentence: "instead of the one specified in the app" 🤔

When an alarm is deleted in app (not the file but the trash icon is clicked when the select alarm window is open) the reference to it is not removed from the defaults in the settings app.

<img width=200 src="https://github.com/user-attachments/assets/d1aa1ea9-353f-4f44-89a2-9e6fa36d3224"><img width=200 src="https://github.com/user-attachments/assets/f76ff89e-937a-4f80-956f-1524a5e3488c">

The uuid at the end means it is still referencing to a file in DP-Storage, but we deleted that when removing the file, so it breaks all alarms using that default sound,
I don't know where this is handled. Removing an alarm should probably also check if its the default alarm and if so reset the app-default alarm to the system-default alarm.   

Is this pull request still able to pull from my branch? Or do I need to create a new one as this is closed?

BlackyHawky commented:
@LaserKaspar 
> Done, any alarm the would not be played should now correctly fallback to the default sound. (Tested on my device)

I tested your commit 09ea077 and now I can't play the default system ringtone (the ringtone named "Default alarm sound").
So this is definitely not the change to make.

What I thought about changing the [isRingtoneUriReadable()](https://github.com/BlackyHawky/Clock/blob/172da7433f44cab3c34df207b211a30f9c3cdc76/app/src/main/java/com/best/deskclock/utils/RingtoneUtils.java#L78) method is for example to check if the file is both in the phone's internal memory and in the protected storage. Maybe this can help?

In any case, you should not remove the condition `if (Settings.System.DEFAULT_ALARM_ALERT_URI.equals(uri))` so that ExoPlayer can play the device's default system ringtone.

> The uuid at the end means it is still referencing to a file in DP-Storage, but we deleted that when removing the file, so it breaks all alarms using that default sound,
I don't know where this is handled. Removing an alarm should probably also check if its the default alarm and if so reset the app-default alarm to the system-default alarm.

Sorry, but I don't see what you're talking about because on my device, the alarm titles don't appear at all like in your screenshots. 😅
Anyway, maybe there is a change to be made in the [removeCustomRingtoneAsync()](https://github.com/BlackyHawky/Clock/blob/172da7433f44cab3c34df207b211a30f9c3cdc76/app/src/main/java/com/best/deskclock/ringtone/RingtonePickerActivity.java#L721) method.

> Is this pull request still able to pull from my branch? Or do I need to create a new one as this is closed?

The best way to manage PRs is to leave the "main" branch identical to mine and create a new branch dedicated to the desired change.
When the PR is validated, a button will appear to delete the branch you created.

---

Edit: I published commit 1644154 which fixes the default ringtone in Alarms settings when it was removed from the ringtone selector in the expanded alarms view; does this improve your point 2?

LaserKaspar commented:
> What I thought about changing the [isRingtoneUriReadable()](https://github.com/BlackyHawky/Clock/blob/172da7433f44cab3c34df207b211a30f9c3cdc76/app/src/main/java/com/best/deskclock/utils/RingtoneUtils.java#L78) method is for example to check if the file is both in the phone's internal memory and in the protected storage. Maybe this can help?

I don't really understand what you mean. The Ringtone is stored as a path to to DP-Storage if supported, so I don't think I can check if its at both locations because the original location of the ringtone is not stored.

Anyways, I updated the code to handle default alarm URIs as well https://github.com/BlackyHawky/Clock/compare/main...LaserKaspar:Clock:directboot-edgecases. It should work now, I am sorry for not testing it enough. It's pretty similar to how it worked before but it also checks if the URI is readable instead of just assuming it for default sounds. I thought the default URI is an actual URI instead of just some kind of "tag" so that's why the change broke it.

> I published commit https://github.com/BlackyHawky/Clock/commit/164415451789f340544091c822fcc86b4112f31c which fixes the default ringtone in Alarms settings when it was removed from the ringtone selector in the expanded alarms view; does this improve your point 2?

Fixed it indeed, they now correctly fallback to system default. I wonder if its possible to delete the system default, but I don't think that's something we need to handle. 😄

> The best way to manage PRs is to leave the "main" branch identical to mine and create a new branch dedicated to the desired change.

Should I open a new pull request for my fix of point 1?

Thanks again for your open-mindedness!

BlackyHawky commented:
@LaserKaspar 
Hi,

Sorry for my late reply.

> Should I open a new pull request for my fix of point 1?

Of course! 😉

> Thanks again for your open-mindedness!

Don't thank me, it's my pleasure. 😉

BlackyHawky commented:
@LaserKaspar 
Just wait until I validate #305 (tonight or tomorrow).
You'll be able to correctly update your main branch and recreate a new one with this change.

BlackyHawky commented:
@LaserKaspar 
> Anyways, I updated the code to handle default alarm URIs as well https://github.com/BlackyHawky/Clock/compare/main...LaserKaspar:Clock:directboot-edgecases. It should work now, I am sorry for not testing it enough. It's pretty similar to how it worked before but it also checks if the URI is readable instead of just assuming it for default sounds. I thought the default URI is an actual URI instead of just some kind of "tag" so that's why the change broke it.

I finally made your change in commit f220776.
So we can conclude that it's now 100% functional?

LaserKaspar commented:
> I finally made your change in commit https://github.com/BlackyHawky/Clock/commit/f220776d2c9e49d67e34a8b3294e2054cd1b430f.

Great!

>  So we can conclude that it's now 100% functional?

I hope so, if any issues arise let me know. From my side it's perfect!
